File: src/interfaces/post.ts
import { prop, Typegoose, ModelType, InstanceType } from 'typegoose';
import IAttachment from './attachment';

export default class IPost extends Typegoose {

	@prop({ required: true })
    schema_version: number;

	@prop({ required: true })
	id: string;

	@prop({ required: true })
	account: string;

	@prop({ required: true })
	content: string;

	@prop({ required: true })
	created_at: number;

	@prop()
	reply_to?: string;

	@prop()
	multimedia?: IAttachment;

	@prop()
	tags?: Array<string>;

	@prop({ required: true })
	likes_count: number;

	@prop({ required: true })
	comments_count: number;

	@prop({ required: true })
	shares_count: number;

	@prop()
	likes?: Array<string>;

	@prop()
	comments?: Array<string>;

	@prop()
	shares?: Array<string>;

	@prop({ required: true })
	visibility: string;

}

File: src/interfaces-internal/post.go
package interfaces_internal

import (
	"github.com/globalsign/mgo"
	"github.com/globalsign/mgo/bson"
	"log"
)

type IPost struct {
	SchemaVersion uint32        `bson:"schema_version"`
	ID            bson.ObjectId `bson:"_id"`
	Account       string        `bson:"account"`
	Content       string        `bson:"content"`
	CreatedAt     int64         `bson:"created_at"`
	ReplyTo       string        `bson:"reply_to"`
	Multimedia    IAttachment   `bson:"multimedia"`
	Tags          []string      `bson:"tags"`
	LikesCount    uint64        `bson:"likes_count"`
	CommentsCount uint64        `bson:"comments_count"`
	SharesCount   uint64        `bson:"shares_count"`
	Likes         []string      `bson:"likes"`
	Comments      []string      `bson:"comments"`
	Shares        []string      `bson:"shares"`
	Visibility    []string      `bson:"visibility"`
}

func InitIPostIndexes(collection *mgo.Collection) {
	err := collection.EnsureIndex(mgo.Index{
		Key: []string{"tags", "_id"},
		Background: true,
	})
	if err != nil {
		log.Fatal(err)
	}
}